File: llamafile/server/www/highlight_make.js
// Copyright 2024 Mozilla Foundation
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

const MAKE_KEYWORDS = new Set([
  '-include',
  '-sinclude',
  'define',
  'else',
  'endef',
  'endif',
  'export',
  'if',
  'ifdef',
  'ifeq',
  'ifndef',
  'ifneq',
  'include',
  'override',
  'private',
  'sinclude',
  'undefine',
  'unexport',
  'vpath',
]);

const MAKE_BUILTINS = new Set([
  'abspath',
  'addprefix',
  'addsuffix',
  'and',
  'basename',
  'call',
  'dir',
  'error',
  'eval',
  'file',
  'filter',
  'filter-out',
  'findstring',
  'firstword',
  'flavor',
  'foreach',
  'if',
  'join',
  'lastword',
  'notdir',
  'or',
  'origin',
  'patsubst',
  'realpath',
  'shell',
  'sort',
  'strip',
  'subst',
  'suffix',
  'value',
  'warning',
  'wildcard',
  'word',
  'wordlist',
  'words',
]);

class HighlightMake extends Highlighter {

  static NORMAL = 0;
  static WORD = 1;
  static COMMENT = 2;
  static COMMENT_BACKSLASH = 3;
  static DOLLAR = 4;
  static DOLLAR2 = 5;
  static VARIABLE = 6;
  static BACKSLASH = 7;

  static is_automatic_variable(c) {
    switch (c) {
    case '@':
    case '%':
    case '<':
    case '?':
    case '^':
    case '+':
    case '|':
    case '*':
      return true;
    default:
      return false;
    }
  }

  constructor(delegate) {
    super(delegate);
    this.word = '';
  }

  feed(input) {
    for (let i = 0; i < input.length; i += this.delta) {
      this.delta = 1;
      let c = input[i];
      switch (this.state) {

      case HighlightMake.NORMAL:
        if (!isascii(c) || isalpha(c) || c == '_' || c == '-' || c == '.') {
          this.state = HighlightMake.WORD;
          this.word += c;
          break;
        } else if (c == '#') {
          this.state = HighlightMake.COMMENT;
          this.push("span", "comment");
          this.append('#');
        } else if (c == '$') {
          this.state = HighlightMake.DOLLAR;
          this.append('$');
        } else if (c == '\\') {
          this.state = HighlightMake.BACKSLASH;
          this.push("span", "escape");
          this.append('\\');
        } else {
          this.append(c);
        }
        break;

      case HighlightMake.BACKSLASH:
        this.append(c);
        this.pop();
        this.state = HighlightMake.NORMAL;
        break;

      case HighlightMake.WORD:
        if (!isascii(c) || isalnum(c) || c == '_' || c == '-' || c == '.') {
          this.word += c;
        } else {
          if (MAKE_KEYWORDS.has(this.word)) {
            this.push("span", "keyword");
            this.append(this.word);
            this.pop();
          } else {
            this.append(this.word);
          }
          this.word = '';
          this.epsilon(HighlightMake.NORMAL);
        }
        break;

      case HighlightMake.DOLLAR:
        if (isdigit(c) || HighlightMake.is_automatic_variable(c)) {
          this.push("span", "var");
          this.append(c);
          this.pop();
          this.state = HighlightMake.NORMAL;
        } else if (c == '$') {
          this.state = HighlightMake.DOLLAR2;
          this.append('$');
        } else if (c == '(') {
          this.state = HighlightMake.VARIABLE;
          this.append('(');
        } else {
          this.epsilon(HighlightMake.NORMAL);
        }
        break;

      case HighlightMake.DOLLAR2:
        if (c == '(') {
          this.state = HighlightMake.VARIABLE;
          this.append('(');
        } else {
          this.epsilon(HighlightMake.NORMAL);
        }
        break;

      case HighlightMake.VARIABLE:
        if (isalnum(c) || //
            c == '%' || //
            c == '*' || //
            c == '+' || //
            c == '-' || //
            c == '.' || //
            c == '<' || //
            c == '?' || //
            c == '@' || //
            c == '_') {
          this.word += c;
        } else if (c == '$' && !this.word) {
          this.state = HighlightMake.DOLLAR;
          this.append('$');
        } else if (c == ')' || //
                   c == ':') {
          this.push("span", "var");
          this.append(this.word);
          this.pop();
          this.word = '';
          this.append(c);
          this.state = HighlightMake.NORMAL;
        } else {
          if (MAKE_BUILTINS.has(this.word)) {
            this.push("span", "builtin");
            this.append(this.word);
            this.pop();
          } else {
            this.append(this.word);
          }
          this.word = '';
          this.epsilon(HighlightMake.NORMAL);
        }
        break;

      case HighlightMake.COMMENT:
        this.append(c);
        if (c == '\n') {
          this.pop();
          this.state = HighlightMake.NORMAL;
        } else if (c == '\\') {
          this.state = HighlightMake.COMMENT_BACKSLASH;
        }
        break;

      case HighlightMake.COMMENT_BACKSLASH:
        this.append(c);
        this.state = HighlightMake.COMMENT;
        break;

      default:
        throw new Error('Invalid state');
      }
    }
  }

  flush() {
    switch (this.state) {
    case HighlightMake.WORD:
      if (MAKE_KEYWORDS.has(this.word)) {
        this.push("span", "keyword");
        this.append(this.word);
        this.pop();
      } else {
        this.append(this.word);
      }
      this.word = '';
      break;
    case HighlightMake.VARIABLE:
      if (MAKE_BUILTINS.has(this.word)) {
        this.push("span", "builtin");
        this.append(this.word);
        this.pop();
      } else {
        this.append(this.word);
      }
      this.word = '';
      break;
    case HighlightMake.COMMENT:
    case HighlightMake.COMMENT_BACKSLASH:
    case HighlightMake.BACKSLASH:
      this.pop();
      break;
    default:
      break;
    }
    this.state = HighlightMake.NORMAL;
    this.delegate.flush();
    this.delta = 1;
  }
}

Highlighter.REGISTRY['make'] = HighlightMake;
Highlighter.REGISTRY['mk'] = HighlightMake;
Highlighter.REGISTRY['gmake'] = HighlightMake;
Highlighter.REGISTRY['makefile'] = HighlightMake;
Highlighter.REGISTRY['gmakefile'] = HighlightMake;

File: llamafile/highlight/highlight_make.cpp
// -*- mode:c++;indent-tabs-mode:nil;c-basic-offset:4;coding:utf-8 -*-
// vi: set et ft=cpp ts=4 sts=4 sw=4 fenc=utf-8 :vi
//
// Copyright 2024 Mozilla Foundation
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "highlight.h"
#include "util.h"
#include <cosmo.h>
#include <ctype.h>

enum
{
    NORMAL,
    WORD,
    COMMENT,
    COMMENT_BACKSLASH,
    DOLLAR,
    DOLLAR2,
    VARIABLE,
    BACKSLASH,
};

static bool
is_automatic_variable(int c)
{
    switch (c) {
    case '@':
    case '%':
    case '<':
    case '?':
    case '^':
    case '+':
    case '|':
    case '*':
        return true;
    default:
        return false;
    }
}

HighlightMake::HighlightMake()
{
}

HighlightMake::~HighlightMake()
{
}

void
HighlightMake::feed(std::string* r, std::string_view input)
{
    for (size_t i = 0; i < input.size(); ++i) {
        wchar_t c;
        int b = input[i] & 255;
        if (!u_) {
            if (b < 0300) {
                c = b;
            } else {
                c_ = ThomPikeByte(b);
                u_ = ThomPikeLen(b) - 1;
                continue;
            }
        } else if (ThomPikeCont(b)) {
            c = c_ = ThomPikeMerge(c_, b);
            if (--u_)
                continue;
        } else {
            u_ = 0;
            c = b;
        }
        if (c == 0xFEFF)
            continue; // utf-8 bom
        switch (t_) {

        Normal:
        case NORMAL:
            if (!isascii(c) || isalpha(c) || c == '_' || c == '-' || c == '.') {
                t_ = WORD;
                lf::append_wchar(&word_, c);
                break;
            } else if (c == '#') {
                t_ = COMMENT;
                *r += HI_COMMENT;
                *r += '#';
            } else if (c == '$') {
                t_ = DOLLAR;
                *r += '$';
            } else if (c == '\\') {
                t_ = BACKSLASH;
                *r += HI_ESCAPE;
                *r += '\\';
            } else {
                lf::append_wchar(r, c);
            }
            break;

        case BACKSLASH:
            lf::append_wchar(r, c);
            *r += HI_RESET;
            t_ = NORMAL;
            break;

        case WORD:
            if (!isascii(c) || isalnum(c) || c == '_' || c == '-' || c == '.') {
                lf::append_wchar(&word_, c);
            } else {
                if (is_keyword_make(word_.data(), word_.size())) {
                    *r += HI_KEYWORD;
                    *r += word_;
                    *r += HI_RESET;
                } else {
                    *r += word_;
                }
                word_.clear();
                t_ = NORMAL;
                goto Normal;
            }
            break;

        case DOLLAR:
            if (isdigit(c) || is_automatic_variable(c)) {
                *r += HI_VAR;
                lf::append_wchar(r, c);
                *r += HI_RESET;
                t_ = NORMAL;
            } else if (c == '$') {
                t_ = DOLLAR2;
                *r += '$';
            } else if (c == '(') {
                t_ = VARIABLE;
                *r += '(';
            } else {
                t_ = NORMAL;
                goto Normal;
            }
            break;

        case DOLLAR2:
            if (c == '(') {
                t_ = VARIABLE;
                *r += '(';
            } else {
                t_ = NORMAL;
                goto Normal;
            }
            break;

        case VARIABLE:
            if (isalnum(c) || //
                c == '%' || //
                c == '*' || //
                c == '+' || //
                c == '-' || //
                c == '.' || //
                c == '<' || //
                c == '?' || //
                c == '@' || //
                c == '_') {
                lf::append_wchar(&word_, c);
            } else if (c == '$' && word_.empty()) {
                t_ = DOLLAR;
                *r += '$';
            } else if (c == ')' || //
                       c == ':') {
                *r += HI_VAR;
                *r += word_;
                *r += HI_RESET;
                word_.clear();
                lf::append_wchar(r, c);
                t_ = NORMAL;
            } else {
                if (is_keyword_make_builtin(word_.data(), word_.size())) {
                    *r += HI_BUILTIN;
                    *r += word_;
                    *r += HI_RESET;
                } else {
                    *r += word_;
                }
                word_.clear();
                t_ = NORMAL;
                goto Normal;
            }
            break;

        case COMMENT:
            lf::append_wchar(r, c);
            if (c == '\n') {
                *r += HI_RESET;
                t_ = NORMAL;
            } else if (c == '\\') {
                t_ = COMMENT_BACKSLASH;
            }
            break;

        case COMMENT_BACKSLASH:
            lf::append_wchar(r, c);
            t_ = COMMENT;
            break;

        default:
            __builtin_unreachable();
        }
    }
}

void
HighlightMake::flush(std::string* r)
{
    switch (t_) {
    case WORD:
        if (is_keyword_make(word_.data(), word_.size())) {
            *r += HI_KEYWORD;
            *r += word_;
            *r += HI_RESET;
        } else {
            *r += word_;
        }
        word_.clear();
        break;
    case VARIABLE:
        if (is_keyword_make_builtin(word_.data(), word_.size())) {
            *r += HI_BUILTIN;
            *r += word_;
            *r += HI_RESET;
        } else {
            *r += word_;
        }
        word_.clear();
        break;
    case COMMENT:
    case COMMENT_BACKSLASH:
    case BACKSLASH:
        *r += HI_RESET;
        break;
    default:
        break;
    }
    c_ = 0;
    u_ = 0;
    t_ = NORMAL;
}
